### ISSUE TITLE
Feature: Word Count as a Data Layer Variable

### ISSUE THREAD

**Issue by ju1** [duracelltomi/gtm4wp#56]
It would be awesome to have the word count in the data layer when the page loads. Even more useful could be "Range of word count", possible values:
- 0-500
- 501-1000
- 1001 - 1500
- 1501 - 2000
- 2001 - 2500 etc.

**Comment by koconder** [comment]
@ju1 might look into this, will see how it affects speed/rendering first. Also in regards to "ranges" this should be handled in the analytics platform (via calculated dimensions) or the tag manager to keep it clean, as ranges might differ for users needs.

**Comment by ju1** [comment]
Thanks for the update. And that's a good idea to calculate ranges on our own.

**Comment by tijevlam** [comment]
Wouldn't it make more sense to calculate the word count directly in GTM? Thats easy with a simple javascript and gives you a lot of control on which sections of content you would like to include in the wordcount.

**Comment by koconder** [comment]
@duracelltomi i have a simple working solution that uses the WP_Query loop. Will add in once the other PR's are cleared so we don't have too many conflicting changes. If there is no harm on performance we can add it in, if there is we can decide if to include with any warnings.

**Comment by duracelltomi** [comment]
Thanks! Two PRs still needs some change as far as I can see but both are very close to be merged i guess :-)

In my experience adding a warning is always the best way: WordPress can be running on various servers with lots of plugins, there will be always cases where such an algorithm will degrade performance.

**Comment by tijevlam** [comment]
@duracelltomi i see your purpose with the plugin. I'm mostly fan of the option to include variables that are available only or more easily on the server side of things. 

I agree it would come in handy when the wordcount would be available straight from this plugin. However, I don't think this should have a significant effect on performance.

I'm eager to see how your solution works @koconder. I've seen some specific wordcount plugins for WP, they all save the wordcount in a new table (or column) in the posts-table. Which would bea great effect solution performancewise i'd say, but would mean a lot of extra implications for this plugin. 

Adding to the wordcount, from a analytist-perspective: I'd suggest we also add imageCount and videoCount. I've made reports and analysis in the past that exposed a real cohersion between reads, scroll depths and number of images/media.

**Comment by duracelltomi** [comment]
having the word count cached in post_meta would be a good idea to minimize impact on performance - but we need to prepare a code that will store this for already existing posts - I guess this can be done while a post is being viewed by a visitor as this would only affect this one page load perhaps

**Comment by duracelltomi** [comment]
I wonder if this can be done just by using [str_word_count()](https://www.php.net/manual/en/function.str-word-count.php)

Perhaps a cleaning would be necessary first:
https://plugins.trac.wordpress.org/browser/wp-word-count/trunk/includes/class-wpwc-functions.php#L56

I wonder if `apply_filters()` with `'the_content'` would be also needed before counting the words?

**Comment by duracelltomi** [comment]
... or we could just use wp-admin/js/word-count.js

**Comment by camassey** [comment]
Hi @duracelltomi - am I right in thinking this hasn't been merge / released yet? Is there (currently) any plan to?

**Comment by duracelltomi** [comment]
No, it has been not merged yet, I want to finish the next version shortly.

Due to covid and being a bit more tired at work, I was adding more vacation and time with my family but I intend to speed up things in the coming weeks

**Comment by koconder** [comment]
@duracelltomi my original suggestion was to bake in support for 3rd party word count plugins, and just read data via post_meta into dataLayer which would be easier to support and maintain. As some sites/themes/plugins could have differences in word count and would be better to focus on pulling in data rather than trying to calculate ourselves.

**Comment by Paktas** [comment]
Each plugin has a slightly different way of counting words. If you embedd content - some fail to recognise HTML, some do, some regard fullstop (.) as word delimiter when it's just a decimal.
Would be best to use some methodology that is at the core of WP to nativelly count words.

**Comment by duracelltomi** [comment]
Thanks everyone for the long discussion. This is implemented in the 2.0 rewrite: a new "Content & engagement data" group of Page variables adds content word count (`pageContentWordCount`) and estimated reading time (`pageReadingTime`) — and, as suggested here, it's an option you enable only where you want it. Ships in 2.0 beta1 — closing as done. Thanks for the patience and the ideas!
